mod/consensus/pkg/cometbft/service/abci.go

```diff
@@ -67,98 +70,96 @@ func (s *Service[LoggerT]) InitChain(
 	)
 
 	// Set the initial height, which will be used to determine if we are
-	// proposing
-	// or processing the first block or not.
+	// proposing or processing the first block or not.
 	s.initialHeight = req.InitialHeight
-	if s.initialHeight == 0 { // If initial height is 0, set it to 1
+	if s.initialHeight == 0 {
 		s.initialHeight = 1
 	}
 
 	// if req.InitialHeight is > 1, then we set the initial version on all
 	// stores
 	if req.InitialHeight > 1 {
-		initHeader.Height = req.InitialHeight
 		if err := s.sm.CommitMultiStore().
 			SetInitialVersion(req.InitialHeight); err != nil {
 			return nil, err
 		}
 	}
 
-	s.setState(execModeFinalize)
+	s.finalizeBlockState = s.resetState()
 
 	defer func() {
 		// InitChain represents the state of the application BEFORE the first
 		// block, i.e. the genesis block. This means that when processing the
 		// app's InitChain handler, the block height is zero by default.
-		// However, after Commit is called
-		// the height needs to reflect the true block height.
-		initHeader.Height = req.InitialHeight
+		// However, after genesis is handled the height needs to reflect
+		// the true block height.
+		initHeader := cmtproto.Header{
+			ChainID: req.ChainId,
+			Time:    req.Time,
+			Height:  req.InitialHeight,
+		}
 		s.finalizeBlockState.SetContext(
 			s.finalizeBlockState.Context().WithBlockHeader(initHeader),
 		)
 	}()
 
-	if s.finalizeBlockState == nil {
-		return nil, errors.New("finalizeBlockState is nil")
-	}
-
-	res, err := s.initChainer(s.finalizeBlockState.Context(), req)
+	resValidators, err := s.initChainer(
+		s.finalizeBlockState.Context(),
+		req.AppStateBytes,
+	)
 	if err != nil {
 		return nil, err
 	}
 
-	if res == nil {
+	if len(resValidators) == 0 {
```

Comment:
this is an hollow check, since we generate res in initChainer (it's not the result of s.Middleware.InitGenesis).
Replaced with what I understand it's the check we want

Reply:
another point is that we may have a stricter check here and enforce that we always have at least a validator?